modules/light-clients/08-wasm/keeper/keeper.go

```diff
@@ -134,3 +144,51 @@ func (k Keeper) storeWasmCode(ctx sdk.Context, code []byte) ([]byte, error) {
 
 	return codeHash, nil
 }
+
+func (k Keeper) migrateContractCode(ctx sdk.Context, clientID string, newCodeHash, migrateMsg []byte) error {
+	wasmClientState, err := k.GetWasmClientState(ctx, clientID)
+	if err != nil {
+		return errorsmod.Wrap(err, "failed to retrieve wasm client state")
+	}
+
+	clientStore := k.clientKeeper.ClientStore(ctx, clientID)
+
+	err = wasmClientState.MigrateContract(ctx, k.cdc, clientID, clientStore, newCodeHash, migrateMsg)
+	if err != nil {
+		return errorsmod.Wrap(err, "contract migration failed")
+	}
+
+	oldCodeHash := wasmClientState.CodeHash
+
+	// client state may be updated by the contract migration
+	wasmClientState, err = k.GetWasmClientState(ctx, clientID)
+	if err != nil {
+		// note that this also ensures that the updated client state is
+		// still a wasm client state
+		return errorsmod.Wrap(err, "failed to retrieve the updated wasm client state")
+	}
+
```

Comment:
I don't think so unless we add this check to every time sudo is called which I might be in favour of. What do you think?

Reply:
It'd be an extra state read, but I think it could be a nice safety check to add after sudo/migrate. It would assert:
* wasm client isn't deleted from store
* it isn't changed to another type
* its code hash only changes via MsgMigrateContract

Happy to open an issue and tackle later. It should be non-API breaking check so no rush to include now. What do you think?

Reply:
https://github.com/cosmos/ibc-go/issues/5011